src/MySqlConnector/Protocol/Serialization/SlicedSurrogatesUtility.cs

```diff
@@ -0,0 +1,81 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace MySqlConnector.Protocol.Serialization
+{
+#if !NET45 && !NET461 && !NET471 && !NETSTANDARD1_3 && !NETSTANDARD2_0 && !NETSTANDARD2_1 && !NETCOREAPP2_1
+
+	internal static class SlicedSurrogatesUtility
+	{
+		public static int GetUtf8ByteCount(ReadOnlySpan<char> span, ref char? surrogatePartBuffer)
+		{
+			var containsSurrogateAtBeginning = BitConverter.IsLittleEndian
+				? char.IsLowSurrogate(span[0])
+				: char.IsHighSurrogate(span[0]);
+			var containsSurrogateAtEnding = BitConverter.IsLittleEndian
+				? char.IsHighSurrogate(span[^1])
+				: char.IsLowSurrogate(span[^1]);
+
+			if (containsSurrogateAtBeginning || containsSurrogateAtEnding)
+			{
+				var startPosition = containsSurrogateAtBeginning ? 1 : 0;
+				var length = containsSurrogateAtEnding ? span.Length - 1 : span.Length;
+
+				var operationalSpan = span[startPosition..length];
+
+				var totalCount = Encoding.UTF8.GetByteCount(operationalSpan);
+
+				if (containsSurrogateAtBeginning)
+				{
+					totalCount += Encoding.UTF8.GetByteCount(new[] { surrogatePartBuffer!.Value, span[0] });
```

Comment:
Seems like a potential bug if the user passes in malformed data starting with an unpaired low-surrogate code unit? (It should write out `U+FFFD`, not crash.)